src/main/kotlin/com/exactpro/th2/codec/configuration/ApplicationContext.kt

```diff
@@ -14,142 +14,45 @@
 import com.exactpro.th2.common.schema.message.MessageRouter
-import com.exactpro.th2.sailfish.utils.IMessageToProtoConverter
-import com.exactpro.th2.sailfish.utils.ProtoToIMessageConverter
-import mu.KotlinLogging
-import org.apache.commons.io.FileUtils
-import org.apache.commons.lang3.BooleanUtils.toBoolean
-import org.apache.commons.lang3.math.NumberUtils.*
-import java.io.File
-import java.net.URLClassLoader
-import java.util.*
 
 class ApplicationContext(
     val commonFactory: CommonFactory,
-    val codec: IExternalCodec,
-    val codecFactory: IExternalCodecFactory,
-    val codecSettings: IExternalCodecSettings,
-    val protoToIMessageConverter: ProtoToIMessageConverter,
-    val messageToProtoConverter: IMessageToProtoConverter,
-    val eventBatchRouter: MessageRouter<EventBatch>?
+    val codecFactory: () -> IPipelineCodec,
+    val eventBatchRouter: MessageRouter<EventBatch>
 ) {
-    fun createDecodeProcessor(type: ProcessorType): AbstractCodecProcessor<RawMessageBatch, MessageBatch> {
-        return when (type) {
-            ProcessorType.CUMULATIVE -> CumulativeDecodeProcessor(codecFactory, codecSettings, messageToProtoConverter)
-            ProcessorType.SEQUENTIAL -> SequentialDecodeProcessor(codecFactory, codecSettings, messageToProtoConverter)
-        }
-    }
+    fun createDecodeProcessor(): DecodeProcessor = DecodeProcessor(codecFactory)
 
     companion object {
-        private const val CODEC_IMPLEMENTATION_PATH = "codec_implementation"
-
-        private val logger = KotlinLogging.logger { }
-
         fun create(configuration: Configuration, commonFactory: CommonFactory): ApplicationContext {
-            val codecFactory = loadFactory(configuration.codecClassName)
+            val codecSettings = checkNotNull(configuration.codecSettings) { "configuration${Configuration::codecSettings} is null" }
+
+            runCatching {
+                load<IPipelineCodec>()
+            }.onFailure {
+                throw IllegalStateException("Failed to load codec", it)
+            }
 
             val eventBatchRouter = commonFactory.eventBatchRouter
-            val codecSettings = createSettings(commonFactory, codecFactory, configuration.codecParameters)
-            val codec = codecFactory.createCodec(codecSettings)
-            val dictionaryType = if (OUTGOING in codecSettings.dictionaryTypes) OUTGOING else MAIN
-            val dictionary = checkNotNull(codecSettings[dictionaryType]) { "Dictionary is not set: $dictionaryType" }
-            val protoConverter = ProtoToIMessageConverter(
-                DefaultMessageFactoryProxy(), dictionary, SailfishURI.unsafeParse(dictionary.namespace)
-            )
-            val iMessageConverter = IMessageToProtoConverter()
+            val dictionary = commonFactory.readDictionary(DictionaryType.MAIN).use(XmlDictionaryStructureLoader()::load)
+
+            val codecFactory = {
+                load<IPipelineCodec>().apply {
```

Comment:
Why can't we load it only onece?

Reply:
Because we need to have separate instances for different threads since `IPipelineCodec` isn't meant to be thread-safe